File: chaincode/node/lib/yjwz.js
'use strict';

const { Contract } = require('fabric-contract-api');

class YjwzContract extends Contract {

    // 初始化 Ledger
    async InitLedger(ctx) {
        console.info('Initialize the ledger with some basic data.');
    }

    // 物资入库操作
    async enter(ctx, createdTime, materialsName, notess, materialNumber) {
        const enterHashValue = this.generateHash(createdTime, materialsName, materialNumber, Date.now().toString());
        
        const enterRecord = {
            createdTime,
            materialsName,
            materialNumber: parseInt(materialNumber),
            notess
        };

        await ctx.stub.putState(enterHashValue, Buffer.from(JSON.stringify(enterRecord)));

        console.info(`New Donation: ${enterRecord.materialsName} created at ${enterRecord.createdTime}.`);
        return enterHashValue;
    }

    // 物资出库操作
    async getOut(ctx, createdTime, materialsName, materialNumber, intention, enterHashValue) {
        const getOutHashValue = this.generateHash(createdTime, materialsName, materialNumber, intention, enterHashValue);

        const getOutRecord = {
            createdTime,
            materialsName,
            materialNumber: parseInt(materialNumber),
            intention,
            parentHash: enterHashValue
        };

        await ctx.stub.putState(getOutHashValue, Buffer.from(JSON.stringify(getOutRecord)));

        console.info(`New Get Out: ${getOutRecord.materialsName} dispatched to ${getOutRecord.intention}.`);
        return getOutHashValue;
    }

    // 查询物资入库记录
    async getEnter(ctx, enterHashValue) {
        const enterRecordBytes = await ctx.stub.getState(enterHashValue);
        if (!enterRecordBytes || enterRecordBytes.length === 0) {
            throw new Error(`Enter record with hash ${enterHashValue} does not exist`);
        }
        return JSON.parse(enterRecordBytes.toString());
    }

    // 查询物资出库记录
    async getGetOut(ctx, getOutHashValue) {
        const getOutRecordBytes = await ctx.stub.getState(getOutHashValue);
        if (!getOutRecordBytes || getOutRecordBytes.length === 0) {
            throw new Error(`Get Out record with hash ${getOutHashValue} does not exist`);
        }
        return JSON.parse(getOutRecordBytes.toString());
    }

    // 生成 Hash 值
    generateHash(...inputs) {
        const crypto = require('crypto');
        return crypto.createHash('sha256').update(inputs.join('')).digest('hex');
    }
}

module.exports = YjwzContract;

File: chaincode/go/yjwz/yjwz.go
package main

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"strconv"
	"time"

	"github.com/hyperledger/fabric-contract-api-go/contractapi"
)

// YjwzContract 物资管理的合约
type YjwzContract struct {
	contractapi.Contract
}

// EnterRecord 物资入库记录
type EnterRecord struct {
	CreatedTime   string `json:"createdTime"`
	MaterialsName string `json:"materialsName"`
	MaterialNumber int    `json:"materialNumber"`
	Notess        string `json:"notess"`
}

// GetOutRecord 物资出库记录
type GetOutRecord struct {
	CreatedTime   string `json:"createdTime"`
	MaterialsName string `json:"materialsName"`
	MaterialNumber int    `json:"materialNumber"`
	Intention     string `json:"intention"`
	ParentHash    string `json:"parentHash"`
}

// InitLedger 初始化 Ledger
func (c *YjwzContract) InitLedger(ctx contractapi.TransactionContextInterface) error {
	fmt.Println("Initialize the ledger with some basic data.")
	return nil
}

// Enter 物资入库操作
func (c *YjwzContract) Enter(ctx contractapi.TransactionContextInterface, createdTime string, materialsName string, notess string, materialNumber string) (string, error) {
	materialNum, err := strconv.Atoi(materialNumber)
	if err != nil {
		return "", fmt.Errorf("materialNumber should be an integer: %v", err)
	}

	enterHashValue := c.GenerateHash(createdTime, materialsName, materialNumber, strconv.FormatInt(time.Now().Unix(), 10))

	enterRecord := EnterRecord{
		CreatedTime:   createdTime,
		MaterialsName: materialsName,
		MaterialNumber: materialNum,
		Notess:        notess,
	}

	enterRecordBytes, err := json.Marshal(enterRecord)
	if err != nil {
		return "", fmt.Errorf("failed to marshal enterRecord: %v", err)
	}

	err = ctx.GetStub().PutState(enterHashValue, enterRecordBytes)
	if err != nil {
		return "", fmt.Errorf("failed to put state: %v", err)
	}

	fmt.Printf("New Material: %s created at %s\n", enterRecord.MaterialsName, enterRecord.CreatedTime)
	return enterHashValue, nil
}

// GetOut 物资出库操作
func (c *YjwzContract) GetOut(ctx contractapi.TransactionContextInterface, createdTime string, materialsName string, materialNumber string, intention string, enterHashValue string) (string, error) {
	materialNum, err := strconv.Atoi(materialNumber)
	if err != nil {
		return "", fmt.Errorf("materialNumber should be an integer: %v", err)
	}

	getOutHashValue := c.GenerateHash(createdTime, materialsName, materialNumber, intention, enterHashValue)

	getOutRecord := GetOutRecord{
		CreatedTime:   createdTime,
		MaterialsName: materialsName,
		MaterialNumber: materialNum,
		Intention:     intention,
		ParentHash:    enterHashValue,
	}

	getOutRecordBytes, err := json.Marshal(getOutRecord)
	if err != nil {
		return "", fmt.Errorf("failed to marshal getOutRecord: %v", err)
	}

	err = ctx.GetStub().PutState(getOutHashValue, getOutRecordBytes)
	if err != nil {
		return "", fmt.Errorf("failed to put state: %v", err)
	}

	fmt.Printf("New Get Out: %s dispatched to %s\n", getOutRecord.MaterialsName, getOutRecord.Intention)
	return getOutHashValue, nil
}

// GetEnter 查询物资入库记录
func (c *YjwzContract) GetEnter(ctx contractapi.TransactionContextInterface, enterHashValue string) (*EnterRecord, error) {
	enterRecordBytes, err := ctx.GetStub().GetState(enterHashValue)
	if err != nil {
		return nil, fmt.Errorf("failed to read from world state: %v", err)
	}
	if enterRecordBytes == nil {
		return nil, fmt.Errorf("enter record with hash %s does not exist", enterHashValue)
	}

	var enterRecord EnterRecord
	err = json.Unmarshal(enterRecordBytes, &enterRecord)
	if err != nil {
		return nil, fmt.Errorf("failed to unmarshal enterRecord: %v", err)
	}

	return &enterRecord, nil
}

// GetGetOut 查询物资出库记录
func (c *YjwzContract) GetGetOut(ctx contractapi.TransactionContextInterface, getOutHashValue string) (*GetOutRecord, error) {
	getOutRecordBytes, err := ctx.GetStub().GetState(getOutHashValue)
	if err != nil {
		return nil, fmt.Errorf("failed to read from world state: %v", err)
	}
	if getOutRecordBytes == nil {
		return nil, fmt.Errorf("get out record with hash %s does not exist", getOutHashValue)
	}

	var getOutRecord GetOutRecord
	err = json.Unmarshal(getOutRecordBytes, &getOutRecord)
	if err != nil {
		return nil, fmt.Errorf("failed to unmarshal getOutRecord: %v", err)
	}

	return &getOutRecord, nil
}

// GenerateHash 生成 Hash 值
func (c *YjwzContract) GenerateHash(inputs ...string) string {
	hasher := sha256.New()
	for _, input := range inputs {
		hasher.Write([]byte(input))
	}
	return hex.EncodeToString(hasher.Sum(nil))
}

func main() {
	chaincode, err := contractapi.NewChaincode(new(YjwzContract))
	if err != nil {
		fmt.Printf("Error create YjwzContract chaincode: %v", err)
		return
	}

	if err := chaincode.Start(); err != nil {
		fmt.Printf("Error starting YjwzContract chaincode: %v", err)
	}
}

